### DISCUSSION TITLE
Multi-step forecasts in Time Series

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1942] [FAQs]
# Suggested Previous Reads

   - **https://github.com/pycaret/pycaret/discussions/1761**
   - [Bias-Variance Tradeoff in Time Series](https://towardsdatascience.com/bias-variance-tradeoff-in-time-series-8434f536387a)

# Resources for this FAQ

- **[Jupyter Notebook](https://nbviewer.ipython.org/github/ngupta23/medium_articles/blob/main/time_series/pycaret/pycaret_ts_multi_step_forecasts.ipynb)** (can be opened in Google Colab)

# Problem Statement

Imagine you work for an airline that is growing and want to forecast how many new planes to procure for the future. One way to do this is to forecast the demand in the future and base your orders on this forecast.

Now let's assume that due to various reasons the lead time to procure new planes is 2 years (from the time of order). In such a case, one does not really care about the forecast accuracy in the next 2 years as much as the forecast accuracy after that, say the 3rd year (since the planes will only be available after 2 years).

Using a simple forecast horizon of 3 years to forecast this will impact the forecast's performance since the metrics for the 3rd year will only be 1/3rd of the weight (with 2/3rd of the weight going to the first 2 years which are not important for this problem statement).

In such cases, a multi-step forecast is helpful as indicated in [Forecasting: Principles and Practices](https://otexts.com/fpp3/tscv.html) where we skip the first few future data points and only use the relevant data points to evaluate the performance of the model.

![image](https://user-images.githubusercontent.com/33585645/145672923-22f07800-5779-4546-b507-c805d0c2b463.png)

# Solution in PyCaret

The `pycaret` time series module now supports multi-step forecasts. Let's see how. For the purpose of this illustration, we will use the famous airline dataset. This data set contains the number of airline passengers from 1949-1960. We will use this dataset to forecast the number of passengers expected in 1963 (i.e. the 3rd year from the end of 1960).

## Get Data

```python
y = get_data("airline", verbose=False)
```

## Setup experiment with a forecast horizon
```python
# fh = 3rd year from the end of the data
# 3rd year is the 25th month to 36th month in the future (note numpy is left inclusive and right exclusive)

exp = TSForecastingExperiment()
exp.setup(data=y, fh=np.arange(25,37), session_id=42)
```

## Check correctness of setup

```python
#### Check Train-Test Split ----
exp.plot_model(plot="train_test_split")
```

![image](https://user-images.githubusercontent.com/33585645/145673289-7452d555-ccd1-4d2d-807d-27fc750dbdef.png)

As we can see, the data is split between train and test and the test data ignores the first 2 years (1958 and 1959) and only looks at the 3rd year (1960) which is what we want to evaluate our problem statement correctly.

```python
#### Check cross-validation setup  ----
exp.plot_model(plot="cv")
```

![image](https://user-images.githubusercontent.com/33585645/145673303-0f8e0ebc-11e1-42b8-9e08-37d7bb783ebc.png)

Also, the cross-validation follows the same pattern for each fold (ignore the first 2 future years for the fold and only look at the 3rd year). This will help us evaluate the problem statement correctly.

## Create an initial model

```python
model = exp.create_model("exp_smooth")
```

![image](https://user-images.githubusercontent.com/33585645/145673329-2fa85b48-1fee-44ca-acc9-4af20ac0a9ba.png)

The results for the 3-fold cross-validation are shown. As we can see the cutoff (last point in training data for each fold) for each fold is as per our cross-validation setup from above.

```python
exp.plot_model(model)
```

![image](https://user-images.githubusercontent.com/33585645/145673353-efc222ee-7b5f-4962-8ad0-3ed154ed64aa.png)

The plot shows the forecast using the model trained on the entire "training" dataset. As expected (from the train-test-split plot), the forecast is only for 1960.

## Finalize model for future forecasts

We can fine-tune these results further, but for the purpose of this exercise, we will assume that we are happy with this. The next step is to finalize the model so that we can make "future" predictions that can be used for ordering the planes. We can use the `finalize_model` function for this which trains the model on the entire dataset (from 1949-1960).

```python
final_model = exp.finalize_model(model)
exp.plot_model(final_model)
```
![image](https://user-images.githubusercontent.com/33585645/145673358-6019a3e5-b5bc-4fba-9170-35446fcd0f71.png)

And there we have it - our forecasts for 1963 (3rd year in the future for this dataset) using a model that has been properly evaluated for our problem statement.

# Conclusion

There are many other industries where this same problem statement exists. Take for example the retail industry or any industry related to supply chain. There is generally a lead time from the time one places an order to the time that one receives it. This evaluation method can be applied to all these cases in order to correctly judge the goodness of a forecasting model.

That's it for now. Happy forecasting!

# Suggested Next Reads

- https://github.com/pycaret/pycaret/discussions/1791
- https://github.com/pycaret/pycaret/discussions/1795

**Comment by kushagrgoyal** [comment]
Is the forecasting performed in a Recursive Multistep way or Direct Multistep way? Is there a way to choose and compare these 2 forecasting methodologies?

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
